clang/lib/CodeGen/CGKitsune.cpp

```diff
@@ -223,156 +223,267 @@ void CodeGenFunction::EmitSpawnStmt(const SpawnStmt &S) {
-      LocalDeclMap.erase(LoopVar);
-      IVDeclMap.insert({LoopVar, OuterLoc}); 
-      QualType type = LoopVar->getType();
-      LValue OuterLV = MakeAddrLValue(OuterLoc, type);
-      RValue OuterRV = EmitLoadOfLValue(OuterLV, DI->getBeginLoc()); 
-      ivs.push_back({LoopVar, OuterRV}); 
-    }
+  // As long as the condition is true, iterate the loop.
+  llvm::BasicBlock *ForBody = createBasicBlock("forall.body");
 
-    // create the detach terminator
-    Builder.CreateDetach(ForBody, Continue.getBlock(), SRStart);  
-
-    // Set up IVs to be copied as firstprivate 
-    llvm::Value *Undef = llvm::UndefValue::get(Int32Ty);
-    AllocaInsertPt = new llvm::BitCastInst(Undef, Int32Ty, "",
-                                               ForBody);
-    EmitBlock(ForBody);
-  } else {
-    // Treat it as a non-zero constant.  Don't even create a new block for the
-    // body, just fall into it.
+  // C99 6.8.5p2/p4: The first substatement is executed if the expression
+  // compares unequal to 0.  The condition must be a scalar type.
+  llvm::Value *BoolCondVal = EvaluateExprAsBool(S.getCond());
+  Builder.CreateCondBr(
+      BoolCondVal, Detach, Sync.getBlock(),
+      createProfileWeightsForLoop(S.getCond(), getProfileCount(S.getBody())));
+
+  if (ExitBlock != LoopExit.getBlock()) {
+    EmitBlock(ExitBlock);
+    EmitBranchThroughCleanup(Sync);
   }
 
+  /////////////////////////////////////////////////////////////////////////////
+  // <KITSUNE>
+
+  EmitBlock(Detach);
+
+  // Extract the DeclStmt from the statement init
+  const DeclStmt *DS = cast<DeclStmt>(S.getInit());
+  
+  // Create threadsafe induction variables before the detach and put them in IVDeclMap
+  for (auto *DI : DS->decls()) 
+    EmitIVLoad(dyn_cast<VarDecl>(DI), IVDeclMap);
+
+  // create the detach terminator
+  Builder.CreateDetach(ForBody, Increment.getBlock(), SRStart);
+  
+  // </KITSUNE>
+  /////////////////////////////////////////////////////////////////////////////
+
+  EmitBlock(ForBody);
 
   incrementProfileCounter(&S);
 
   {
     // Create a separate cleanup scope for the body, in case it is not
     // a compound statement.
     RunCleanupsScope BodyScope(*this);
-    if(S.getCond()){
-      for (auto &ivp : ivs){
-        auto* LoopVar = ivp.first; 
-        auto OuterRV = ivp.second; 
-        AutoVarEmission LVEmission = EmitAutoVarAlloca(*LoopVar);
-        EmitAutoVarCleanups(LVEmission);
-        QualType type = LoopVar->getType();
-        Address Loc = LVEmission.getObjectAddress(*this);
-        LValue LV = MakeAddrLValue(Loc, type);	
-        LV.setNonGC(true);
-
-        EmitStoreThroughLValue(OuterRV, LV, true);
-      }
-      EmitStmt(S.getBody());
-    }
-  }
 
-  if(S.getCond()){
-    auto tmp = AllocaInsertPt; 
-    AllocaInsertPt = OldAllocaInsertPt; 
-    tmp->removeFromParent(); 
+    ///////////////////////////////////////////////////////////////////////////
+    // <KITSUNE>
```

Comment:
Just a nit -- the entire source file is for kitsune so the purpose of the "KITSUNE" comments blocks isn't that clear to me.  Are they serving as a marker for a more specific purpose that this comment could provide more info on instead?